src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/mode-base-server/mode-base-server.h>
+#include <app/clusters/on-off-server/on-off-server.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <platform/DiagnosticDataProvider.h>
+
+// using namespace std;
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using chip::Protocols::InteractionModel::Status;
+using BootReasonType = GeneralDiagnostics::BootReasonEnum;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+std::map<uint32_t, Instance *> Instance::ModeBaseAliasesInstanceMap;
+
+bool Instance::HasFeature(Feature feature) const
+{
+    return (mFeature & to_underlying(feature)) != 0;
+}
+
+bool Instance::isAliasCluster() const
+{
+    for (unsigned int AliasedCluster : AliasedClusters)
+    {
+        if (mClusterId == AliasedCluster)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+CHIP_ERROR Instance::Init()
+{
+    // Check that the cluster ID given is a valid mode base alias cluster ID.
+    if (!isAliasCluster())
+    {
+        ChipLogError(Zcl, "ModeBase: The cluster with ID %lu is not a mode base alias.", long(mClusterId));
+        return CHIP_ERROR_INVALID_ARGUMENT; // todo is this the correct error?
+    }
+
+    // Check if the cluster has been selected in zap
+    if (!emberAfContainsServer(mEndpointId, mClusterId))
+    {
+        ChipLogError(Zcl, "ModeBase: The cluster with ID %lu was not enabled in zap.", long(mClusterId));
+        return CHIP_ERROR_INVALID_ARGUMENT; // todo is this the correct error?
+    }
+
+    ReturnErrorOnFailure(chip::app::InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(AppInit());
+
+    ModeBaseAliasesInstanceMap[mClusterId] = this;
+
+    // StartUp behavior relies on CurrentMode StartUpMode attributes being non-volatile.
+    if (!emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::CurrentMode::Id) &&
+        !emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::StartUpMode::Id))
+    {
+        // Read the StartUpMode attribute and set the CurrentMode attribute
```

Comment:
```suggestion
        // Read the StartUpMode attribute and set the CurrentMode attribute. From spec:
```

Reply:
I do not understand this suggestion. Are you requesting adding a reference to the part in the spec that specifies this?

Reply:
Since there's  a spec quote, it should be introduced as such, as opposed to your own words.

Reply:
Ah. I have rewritten the Init function to reduce the dependency on the OnOff cluster and in doing so I had added the spec text.